src/index.js

```diff
@@ -245,64 +235,72 @@ function replaceName(filename, name) {
 	);
 }
 
-function getMain({ options, entry, format }) {
+function getMain({ options, entries, format }) {
 	const { pkg } = options;
 	const pkgMain = options['pkg-main'];
 
-	if (!pkgMain) {
-		return options.output;
-	}
+	const inputs = {};
+	entries.forEach(entry => {
+		let mainNoExtension = options.output;
+
+		if (!pkgMain) {
+			inputs[basename(mainNoExtension, extname(mainNoExtension))] = entry;
+			return;
+		}
+
+		let name;
+		// when we have multiple entries we look for the index file and mark that as our main entry
+		if (entries.length > 1) {
+			name = entry.match(/([\\/])index(\.(umd|cjs|es|m))?\.(mjs|[tj]sx?)$/)
+				? mainNoExtension
+				: entry;
+		} else {
+			name = mainNoExtension;
+		}
 
-	let mainNoExtension = options.output;
-	if (options.multipleEntries) {
-		let name = entry.match(/([\\/])index(\.(umd|cjs|es|m))?\.(mjs|[tj]sx?)$/)
-			? mainNoExtension
-			: entry;
 		mainNoExtension = resolve(dirname(mainNoExtension), basename(name));
-	}
-	mainNoExtension = mainNoExtension.replace(
-		/(\.(umd|cjs|es|m))?\.(mjs|[tj]sx?)$/,
-		'',
-	);
 
-	const mainsByFormat = {};
+		mainNoExtension = mainNoExtension.replace(
+			/(\.(umd|cjs|es|m))?\.(mjs|[tj]sx?)$/,
+			'',
+		);
 
-	mainsByFormat.es = replaceName(
-		pkg.module && !pkg.module.match(/src\//)
-			? pkg.module
-			: pkg['jsnext:main'] || 'x.esm.js',
-		mainNoExtension,
-	);
-	mainsByFormat.modern = replaceName(
-		(pkg.syntax && pkg.syntax.esmodules) || pkg.esmodule || 'x.modern.js',
-		mainNoExtension,
-	);
-	mainsByFormat.cjs = replaceName(pkg['cjs:main'] || 'x.js', mainNoExtension);
-	mainsByFormat.umd = replaceName(
-		pkg['umd:main'] || pkg.unpkg || 'x.umd.js',
-		mainNoExtension,
-	);
+		const mainsByFormat = {};
+
+		mainsByFormat.es = replaceName(
+			pkg.module && !pkg.module.match(/src\//)
+				? pkg.module
+				: pkg['jsnext:main'] || 'x.esm.js',
+			mainNoExtension,
+		);
+		mainsByFormat.modern = replaceName(
+			(pkg.syntax && pkg.syntax.esmodules) || pkg.esmodule || 'x.modern.js',
+			mainNoExtension,
+		);
+		mainsByFormat.cjs = replaceName(pkg['cjs:main'] || 'x.js', mainNoExtension);
+		mainsByFormat.umd = replaceName(
+			pkg['umd:main'] || pkg.unpkg || 'x.umd.js',
+			mainNoExtension,
+		);
+
+		const file = mainsByFormat[format] || mainsByFormat.cjs;
+		inputs[basename(file, extname(file))] = entry;
+	});
 
-	return mainsByFormat[format] || mainsByFormat.cjs;
+	return inputs;
```

Comment:
This needs some cleanup but this should be done in a separate PR where we decouple more.